### PR TITLE
Use word wrapping rather than character wrapping

### PR DESCRIPTION
**Pull request by stephenfin**
Currently, the output diagrams will wrap words at a character boundary
rather than a word boundary. This results in some rather ugly diagrams.
Change this.

Note that this patch was previously submitted against the BitBucket
repo [1]. That went unmerged so we've been carrying a version of this in
OpenStack Nova for a few years [2]. It's pretty battle tested.

[1] https://bitbucket.org/blockdiag/blockdiag/pull-requests/16/
[2] https://github.com/openstack/nova/blob/20.0.0/doc/source/conf.py#L170-L233

### CODE REVIEW

File: src/blockdiag/imagedraw/textfolder.py [hunk 1]
```diff
@@ -13,38 +13,40 @@
...
+    Every line will be stripped. If text includes characters "\n\n", treat
+    as line separator. Ignore '\n' to allow line wrapping.
     """
-    string = re.sub(r'^\s*', '', string)
-    string = re.sub(r'\s*$', '', string)
-    string = re.sub(r'\xa5', '\\\\', string)
-    string = re.sub('(\\\\){2}', '\x00', string)
-    string = re.sub('\\\\n', '\n', string)
-    for line in string.splitlines():
-        yield re.sub('\x00', '\\\\', line).strip()
```
Comment (tk0miya) [inline src/blockdiag/imagedraw/textfolder.py:32]: Why did you remove these code? I can't accept such breaking change.

File: src/blockdiag/imagedraw/textfolder.py [hunk 2]
```diff
@@ -13,38 +13,40 @@
...
+def splittext(metrics, text, bound, measure='width'):
+    folded = [' ']
+    for word in text.split():
+        # Try appending the word to the last line
+        tryline = ' '.join([folded[-1], word]).strip()
+        textsize = metrics.textsize(tryline)
+
+        if getattr(textsize, measure) > bound:
+            # Start a new line. Appends `word` even if > bound.
+            folded.append(word)
```
Comment (tk0miya) [inline src/blockdiag/imagedraw/textfolder.py:47]: It seems empty string is remaining on `folded` as a line if first word is larger than bound. is it intended?

Comment (tk0miya) [inline src/blockdiag/imagedraw/textfolder.py:47]: This change causes word goes out-of-box. I can accept folding by word basically. But this is not.

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: src/blockdiag/imagedraw/textfolder.py (hunk 1 above)
Comment (stephenfin) [reply src/blockdiag/imagedraw/textfolder.py:32]: This was mostly accidental. However, I've no idea what this is doing. Could you help me understand so I can adjust my change to suit?

Comment (stephenfin) [reply src/blockdiag/imagedraw/textfolder.py:32]: I've groked what most of it's doing, but I have some questions.

    string = re.sub(r'^\s*', '', string)

Removes leading whitespace. Handled by `strip()`

    string = re.sub(r'\s*$', '', string)

Removes trailing whitespace. Handled by `strip()`

    string = re.sub(r'\xa5', '\\\\', string)

Replaces the Yen symbol with `\\\\`. I don't know why this is happening. Do you?

    string = re.sub('(\\\\){2}', '\x00', string)

Replaces `\\\\` with `\x00` (the NULL character). Again, I don't know why this is done.

    string = re.sub('\\\\n', '\n', string)

Replaces an escaped newline with an actual newline. Handled by `splitlines()`?

    yield re.sub('\x00', '\\\\', line).strip()

Replaces the null character added previously with `\\\\`. Again, why?

**Pull request closed without merging**